File: Model/oggettoceleste.cpp
#include "oggettoceleste.h"
#include <algorithm>
#include <math.h>
#include <typeinfo>

const double OggettoCeleste::pi= 3.141592653589793;

OggettoCeleste::OggettoCeleste(double r, double ts, double dm, unsigned int e): raggio(r),
    temperaturaSuperficiale(ts), densitaMedia(dm), eta(e){
    if(r <= 0 || dm <= 0){
        throw EccInput("Raggio e temperatura superficiale devono essere > 0");
    }
}

OggettoCeleste::~OggettoCeleste(){}

double OggettoCeleste::getRaggio() const{
    return raggio;
}

double OggettoCeleste::getTemp() const{
    return temperaturaSuperficiale;
}

double OggettoCeleste::getDens() const{
    return densitaMedia;
}

unsigned int OggettoCeleste::getEta() const{
    return eta;
}


vector<const OggettoCeleste*> OggettoCeleste::ordinaPer(vector<const OggettoCeleste*>& v, char condizione){

    std::sort(v.begin(), v.end(),
              [condizione](const OggettoCeleste* primo,const OggettoCeleste* secondo){

                if(condizione == 'v') return *primo < *secondo;
                else return primo->eta < secondo->eta;
            }
    );

    return v;
}

double OggettoCeleste::Volume()const{
    return (pi/6)*pow(raggio*2, 3);
}

double OggettoCeleste::rapportaVolume(const OggettoCeleste& ogg) const{
    return Volume()/ogg.Volume();
}

double OggettoCeleste::Massa() const{
    return densitaMedia*Volume();
}

double OggettoCeleste::Superficie() const{
    return 4*pi*pow(raggio,2);
}

void OggettoCeleste::fusione(const OggettoCeleste& ogg){
    raggio+=ogg.raggio;
    densitaMedia=(densitaMedia+ogg.densitaMedia)/2;
    temperaturaSuperficiale=(temperaturaSuperficiale+ogg.temperaturaSuperficiale)/2;
    eta=0;
}

bool OggettoCeleste::operator>(const OggettoCeleste& ogg) const{
    return raggio > ogg.raggio;
}

bool OggettoCeleste::operator<(const OggettoCeleste& ogg) const{
    return raggio < ogg.raggio;
}

File: Java/src/astrokalk/OggettoCeleste.java
package astrokalk;

import java.util.Comparator;
import java.util.Vector;

import static java.lang.Math.pow;

public abstract class OggettoCeleste {

	private double raggio;
	private double temperaturaSuperficiale;
	private double densitaMedia;
	private int eta;

	protected  static double pi =  3.141592653589793;

	public OggettoCeleste(double r, double ts, double d, int e) throws EccInput {
		if (r <= 0 || d <= 0 || e < 0) {
			throw new EccInput("Raggio e densità devono essere > 0, età >=0");
		}
		raggio = r;
		temperaturaSuperficiale = ts;
		densitaMedia = d;
		eta = e;
	}

	public OggettoCeleste(OggettoCeleste obj){
		raggio = obj.raggio;
		temperaturaSuperficiale = obj.temperaturaSuperficiale;
		densitaMedia = obj.densitaMedia;
		eta = obj.eta;
	}

	public double getRaggio() {
		return raggio;
	}

	public double getTemp() {
		return temperaturaSuperficiale;
	}

	public double getDens() {
		return densitaMedia;
	}

	public int getEta() {
		return eta;
	}

	public static Vector<OggettoCeleste> ordinaPer(Vector<OggettoCeleste> v, char condizione) {

		Vector<OggettoCeleste> aux = new Vector<>(v);

		aux.sort(Comparator.comparing(primo -> {
			if(condizione == 'v')  return Double.valueOf(primo.raggio);
			else return Double.valueOf(primo.eta);
		}));
		return aux;

	}

	public double volume(){
		return (pi/6)*pow(raggio*2, 3);
	}

	public double rapportaVolume(OggettoCeleste ogg){
		return volume()/ogg.volume();
	}

	public double massa(){
		return densitaMedia*volume();
	}

	public double superficie(){
		return 4*pi*pow(raggio,2);
	}

	public void fusione(OggettoCeleste ogg){
		raggio+=ogg.raggio;
		densitaMedia=(densitaMedia+ogg.densitaMedia)/2;
		temperaturaSuperficiale=(temperaturaSuperficiale+ogg.temperaturaSuperficiale)/2;
		eta=0;
	}

	public boolean maggioreDi(OggettoCeleste obj) {
		return raggio > obj.raggio;
	}

	public boolean minoreDi(OggettoCeleste obj) {
		return raggio < obj.raggio;
	}

	@Override
	public String toString() {
		return "Raggio: " + raggio + " km, temp. super.: " + temperaturaSuperficiale + " C°, densità media: " + densitaMedia + " g/cm^3 età: " + eta + " milioni di anni ";
	}
}
